river/cluster/evostream.py

```diff
@@ -0,0 +1,424 @@
+            for i in self._cluster_population
+        }
+
+        p_1, p_2 = self._roulette_wheel_selection(self._fitness_cluster_population)
+        o_1, o_2 = self._binary_crossover(p_1, p_2)
+        delta = random.uniform(0, 1)
+
+        def mutate(g):
+            for i in g:
+                if g[i] == 0:
+                    g[i] = 2 * delta if random.uniform(0, 1) < 0.5 else -2 * delta
+                else:
+                    g[i] = (
+                        2 * delta * g[i]
+                        if random.uniform(0, 1) < 0.5
+                        else -2 * delta * g[i]
+                    )
+                return g
+
+        if delta < self.mutation_probability:
+            for i in o_1:
+                o_1[i].center = mutate(o_1[i].center)
+                o_2[i].center = mutate(o_2[i].center)
+
+        fitness_o_1 = self._fitness(self._evolution_batch, o_1)
+        fitness_o_2 = self._fitness(self._evolution_batch, o_2)
+
+        if fitness_o_1 > min(self._fitness_cluster_population.values()):
+            argmin_1 = np.argmin(list(self._fitness_cluster_population.values()))
+            self._cluster_population[argmin_1] = o_1
+            self._fitness_cluster_population[argmin_1] = fitness_o_1
+
+        if fitness_o_2 > min(self._fitness_cluster_population.values()):
+            argmin_2 = np.argmin(list(self._fitness_cluster_population.values()))
+            self._cluster_population[argmin_2] = o_2
+            self._fitness_cluster_population[argmin_2] = fitness_o_2
+
+    def _merge_micro_clusters(self, micro_clusters):
+        merged_status = {i: False for i in micro_clusters}
+        merged_mcs = {}
+        n_micro_cluster = len(micro_clusters)
+        count = 0
+        for i in micro_clusters:
+            if merged_status[i]:
+                continue
+            else:
+                merged_mcs[count] = micro_clusters[i]
+                for j in range(i + 1, n_micro_cluster):
+                    if (not merged_status[j]) and self._distance(
+                        merged_mcs[count].center, micro_clusters[j].center
+                    ) <= self.radius:
+                        merged_mcs[count].merge(micro_clusters[j])
+                        merged_status[j] = True
+                count += 1
+        return merged_mcs
+
+    def _cleanup(self):
+        # Algorithm 2 of Michael Hahsler and Matthew Bolanos: Cleanup process to remove
+        # inactive clusters and shared density entries from memory
+
+        for i in list(self._micro_clusters):
+            self._micro_clusters[i].weight = self._micro_clusters[i].weight * (
+                2
+                ** (
+                    -self.decay_rate
+                    * (self.time_stamp - self._micro_clusters[i].last_update)
+                )
+            )
+            if self._micro_clusters[i].weight < 2 ** (
+                -self.decay_rate * self.cleanup_interval
+            ):
+                self._micro_clusters.pop(i)
```

Comment:
If the returned value will not be used and `i` is expected to be in the dict, why not using `del` instead?

Reply:
`del` used (honestly I did not pay attention to this at all while coding)!